asaps/models.py

```diff
@@ -0,0 +1,132 @@
+from asnake.client import ASnakeClient
+import importlib
+import json
+import attr
+
+
+@attr.s
+class Client:
+    authclient = attr.ib
+
+    def createclient(secfile):
+        """Select secrets.py file for the appropriate instance."""
+        secfileExists = importlib.util.find_spec(secfile)
+        if secfileExists is not None:
+            secrets = __import__(secfile)
+        else:
+            secrets = __import__('secrets')
+        print('Editing ' + secfile + ' ' + secrets.baseURL)
+        authclient = ASnakeClient(baseurl=secrets.baseURL,
+                                  username=secrets.user,
+                                  password=secrets.password)
+        authclient.authorize()
+        setattr(Client, 'authclient', authclient)
+
+    def getrecord(uri):
+        """Retrieve an individual record."""
+        record = Client.authclient.get(uri).json()
+        print(uri)
+        recType = record['jsonmodel_type']
+
+        # record type
+        if recType == 'resource':
+            recObj = Resource.classpop(record, Resource)
```

Comment:
Our data model classes should ideally not have any complicated logic for creating instances of themselves. In this case, there's some logic involved in pulling data out of the JSON and using that to create the class. The data model classes shouldn't know anything about this JSON record as we'd like to keep that format specific concern further up the in the app.

The first thing I'd do is get rid of the `basepop` and `classpop` methods on the `BaseRecord` class. All it will be is a class with a bunch of attributes. Then move each of the keylist attributes in the other classes to just plain attributes, for example:

```python
@attr.s
class Resource(BaseRecord):
    related_accessions = attr.ib()
    tree = attr.ib()
    related_accessions = attr.ib()
    tree = attr.ib()
```

There are a couple ways we can address the problem of getting the data from the JSON into the class attributes. The first one is going to be more typing, but it's also more straightforward and probably what I'd do if I were using plain old Python classes. As a partial example just to get an idea, you could do something like:

```python
if recType == "resource":
    rec = Resource(uri=record['uri'], title=record['title'], ... # etc)
```

This gives you a lot more control over mapping which fields in the JSON record map to which fields in the class object, but it's obviously more work. Since we're using attrs we can probably do something a bit fancier here:

```python
import operator

f = operator.attrgetter("name")

# <snip>

if recType == "resource":
    fields = [f(field) for field in attr.fields(Resource)]
    kwargs = {k: v for k, v in record.items() if k in fields}
    rec = Resource(objtype=record['type'], **kwargs)
elif recType == "accession":
    fields = [f(field) for field in attr.fields(Accession)]
    kwargs = {k: v for k, v in record.items() if k in fields}
    rec = Accession(objtype=record['type'], **kwargs)
...
```